### PR TITLE
test: add simulation tests

### PR DESCRIPTION
**Pull request by kingpinXD**
# Description

Add simulation tests to the node repository 
Use the following commands to run 
```
make test-sim-nondeterminism
make test-sim-fullappsimulation

```
Closes : https://github.com/zeta-chain/node/issues/2960

The simulation tests would be added into the the CI as part of this issue : https://github.com/zeta-chain/node/issues/2960

# How Has This Been Tested?

- [ ] Tested CCTX in localnet
- [ ] Tested in development environment
- [ ] Go unit tests
- [ ] Go integration tests
- [ ] Tested via GitHub Actions

## Release Notes

- **New Features**
  - Introduced a new module management system for enhanced modularity within the application.
  - Added support for simulation tests, including various test targets for different simulation scenarios.

- **Bug Fixes**
  - Resolved issues with voting mechanisms and transaction processing on EVM chains.

- **Documentation**
  - Updated changelog to reflect new features, refactors, and improvements across multiple versions.

- **Chores**
  - Added a new Makefile to facilitate simulation test execution. 

- **Tests**
  - Implemented a comprehensive testing suite for application simulations, focusing on determinism and state export functionality.

### CODE REVIEW

File: tests/simulation/sim_test.go [hunk 1]
```diff
@@ -0,0 +1,231 @@
+package simulation_test
```
Comment (kingpinXD) [inline tests/simulation/sim_test.go:1]: I created a new folder tests. 
Planning to move towards creating subfolders under this, in the future
- simulation 
- e2e 
- ...

File: tests/simulation/sim_test.go [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+// AppChainID hardcoded chainID for simulation
+
+func init() {
+	simutils.GetSimulatorFlags()
+}
+
+const (
+	SimAppChainID  = "simulation_777-1"
+	SimBlockMaxGas = 815000000
+	SimDBBackend   = "goleveldb"
```
Comment (kingpinXD) [inline tests/simulation/sim_test.go:33]: simulation helper funtions only supports goleveldb and memdb backed for now
We can consider writing out own setup function to use pebbleDB if needed

**Review by lumtis — COMMENTED** [review]
Anyway we can add instruction how to run?
Can be added in `docs/development` with a link in the readme

**Comment by kingpinXD** [comment]
> Anyway we can add instruction how to run? Can be added in `docs/development` with a link in the readme

c1db4cf4cedc9086d3e3c67a820aea587149f513 
I added a simple doc will add more information to it about how the tests work , and the different parameters to adjust

File: docs/development/SIMULATION_TESTING.md [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+## Running the simulation tests
+
+The simulation tests can be run with the following command:
+```bash
+
+make -f sims.mk  test-sim-nondeterminism
+make -f sims.mk  test-sim-fullappsimulation
+make -f sims.mk  test-sim-multi-seed-long
+make -f sims.mk  test-sim-multi-seed-short
```
Comment (lumtis) [inline docs/development/SIMULATION_TESTING.md:12]: Can we precise the different between those?

File: changelog.md [hunk 4]
```diff
@@ -36,6 +36,7 @@
 * [2874](https://github.com/zeta-chain/node/pull/2874) - add support for multiple runs for precompile tests
 * [2895](https://github.com/zeta-chain/node/pull/2895) - add e2e test for bitcoin deposit and call
 * [2894](https://github.com/zeta-chain/node/pull/2894) - increase gas limit for TSS vote tx
+* [2947](https://github.com/zeta-chain/node/pull/2947/) - add simulation tests
```
Comment (lumtis) [inline changelog.md:39]: I would call it more, initialize simulation tests as we will need to add crosschain simulation in the future

File: sims.mk [hunk 5]
```diff
@@ -0,0 +1,48 @@
+#!/usr/bin/make -f
```
Comment (lumtis) [inline sims.mk:1]: Why creating a dedicated makefile for simulation tests and not reusing Makefile?

Our default Makefile is already general purpose

File: sims.mk [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+
+define run-sim-test
+	@echo "Running $(1)..."
+	@go test -mod=readonly $(SIMAPP) -run $(2) -Enabled=true \
+		-NumBlocks=$(3) -BlockSize=$(4) -Commit=true -Period=0 -v -timeout $(5)
+endef
+
+test-sim-nondeterminism:
+	$(call run-sim-test,"non-determinism test",TestAppStateDeterminism,100,200,2h)
```
Comment (lumtis) [inline sims.mk:28]: Does it mean it run 200 batch? Why do we need that many?

Comment (kingpinXD) [reply sims.mk:28]: While testing, I noticed that sometimes the test only fails for larger batch sizes; this is not a finalized number and can be tweaked if needed.

**Review by lumtis — COMMENTED** [review]
Can we tweak staking params?

The staking hook issue was not ported on `develop`: https://github.com/zeta-chain/node/pull/2968

I was curious if setting a unbonding time of 10 seconds could lead to catching the error in the simulation tests

File: sims.mk (hunk 6 above)
Comment (lumtis) [reply sims.mk:28]: Will need to be decreased for CI integration

File: sims.mk [hunk 7]
```diff
@@ -0,0 +1,48 @@
+#!/usr/bin/make -f
+
+########################################
+### Simulations
+
+BINDIR ?= $(GOPATH)/bin
+SIMAPP = ./tests/simulation
+
+
+# Run sim is a cosmos tool which helps us to run multiple simulations in parallel.
```
[the remark this answers is not in the document]
Comment (kingpinXD) [reply sims.mk:10]: The runsim tool, allows multiple threads to run the same simulation at one time, Its not mandatory but good to have. I checked some other repos , and they seem tobe using it (Gaia and Chronos)

File: app/app.go [hunk 8]
```diff
@@ -647,11 +619,14 @@ func New(
...
 		authAddr,
 	)
+
 	app.GovKeeper = *govKeeper.SetHooks(
 		govtypes.NewMultiGovHooks(
 		// register governance hooks
 		),
 	)
+	// Set legacy router for backwards compatibility with gov v1beta1
+	app.GovKeeper.SetLegacyRouter(govRouter)
```
[the remark this answers is not in the document]
Comment (kingpinXD) [reply app/app.go:629]: The simulation tests for the gov module add legacy messages.I don't see an issue with us adding support for that on our application as well.

**Comment by kingpinXD** [comment]
> Can we tweak staking params?
> 
> The staking hook issue was not ported on `develop`: #2968
> 
> I was curious if setting a unbonding time of 10 seconds could lead to catching the error in the simulation tests

Tried it out , I wasn't able to produce the error . 
The unbond amount is randomly selected , 
```
		unbondAmt, err := simtypes.RandPositiveInt(r, totalBond)
		if err != nil {
			return simtypes.NoOpMsg(types.ModuleName, types.TypeMsgUndelegate, "invalid unbond amount"), nil, err
		}

		if unbondAmt.IsZero() {
			return simtypes.NoOpMsg(types.ModuleName, types.TypeMsgUndelegate, "unbond amount is zero"), nil, nil
		}

		msg := types.NewMsgUndelegate(
			delAddr, valAddr, sdk.NewCoin(k.BondDenom(ctx), unbondAmt),
		)

```

Which might be the reason that it is not hitting the edge case which caused the panic in our tests

File: app/app.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (lumtis) [reply app/app.go:629]: Let's remove it, it's out of scope for this PR

File: app/app.go [hunk 9]
```diff
@@ -737,6 +712,8 @@ func New(
 		authzmodule.NewAppModule(appCodec, app.AuthzKeeper, app.AccountKeeper, app.BankKeeper, app.interfaceRegistry),
 	)
 
+	app.ModuleBasics = ModuleBasics
```
Comment (lumtis) [inline app/app.go:715]: Why do we need this change by setting ModuleBasics in app?

File: app/app.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (kingpinXD) [reply app/app.go:629]: I can remove the gov module from the sim tests for now in that case , as it includes legacy proposals in its simulation

File: app/app.go (hunk 9 above)
Comment (kingpinXD) [reply app/app.go:715]: The default genesis function is now part of module basics , which we use in the simulation tests 
// DefaultGenesis provides default genesis information for all modules
```
func (bm BasicManager) DefaultGenesis(cdc codec.JSONCodec) map[string]json.RawMessage {
	genesis := make(map[string]json.RawMessage)
	for _, b := range bm {
		if mod, ok := b.(HasGenesisBasics); ok {
			genesis[b.Name()] = mod.DefaultGenesis(cdc)
		}
	}

	return genesis
}
```

File: app/modules.go [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+	lightclienttypes "github.com/zeta-chain/node/x/lightclient/types"
+	observermodule "github.com/zeta-chain/node/x/observer"
+	observertypes "github.com/zeta-chain/node/x/observer/types"
+)
+
+// ModuleBasics defines the module BasicManager that is in charge of setting up basic,
+// non-dependant module elements, such as codec registration
+// and genesis verification.
+
+// TODO: Use app.mm to create the basic manager instead
```
Comment (lumtis) [inline app/modules.go:60]: Do we have an issue? Maybe we can add more details on this?

File: app/app.go (hunk 9 above)
Comment (lumtis) [reply app/app.go:715]: Do we still need `ModuleBasics` as a public variable?

Comment (kingpinXD) [reply app/app.go]: you are correct, we dont 
f8c095d8b9820bc8de07ac0e29c30dce0f991d8c

**Review by lumtis — APPROVED** [review]

File: Makefile [hunk 11]
```diff
@@ -345,6 +345,61 @@ start-upgrade-import-mainnet-test: zetanode-upgrade
...
+$(BINDIR)/runsim:
+	@echo 'Installing runsim...'
+	@TEMP_DIR=$$(mktemp -d) && \
+	cd $$TEMP_DIR && \
+	go install github.com/cosmos/tools/cmd/runsim@v1.0.0 && \
+	rm -rf $$TEMP_DIR || (echo 'Failed to install runsim' && exit 1)
+	@echo 'runsim installed successfully'
+
+
+# Configuaration parameters for simulation tests
```
Comment (fbac) [inline Makefile:368]:
```suggestion
# Configuration parameters for simulation tests
```

File: changelog.md [hunk 12]
```diff
@@ -36,6 +36,7 @@
 * [2874](https://github.com/zeta-chain/node/pull/2874) - add support for multiple runs for precompile tests
 * [2895](https://github.com/zeta-chain/node/pull/2895) - add e2e test for bitcoin deposit and call
 * [2894](https://github.com/zeta-chain/node/pull/2894) - increase gas limit for TSS vote tx
+* [2947](https://github.com/zeta-chain/node/pull/2947/) - initialize simulation tests
```
Comment (fbac) [inline changelog.md:39]:
```suggestion
* [2894](https://github.com/zeta-chain/node/pull/2894) - increase gas limit for TSS vote tx
* [2947](https://github.com/zeta-chain/node/pull/2947) - initialize simulation tests
```

File: docs/development/SIMULATION_TESTING.md [hunk 13]
```diff
@@ -0,0 +1,37 @@
+# Zetachain simulation testing
```
Comment (fbac) [inline docs/development/SIMULATION_TESTING.md:1]: Markdown version formatted by a linter and some minor corrections:
----BEGIN
# Zetachain simulation testing

## Overview

The blockchain simulation tests how the blockchain application would behave under real life circumstances by generating and sending randomized messages.
The goal of this is to detect and debug failures that could halt a live chain, by providing logs and statistics about the operations run by the simulator as well as exporting the latest application state.

## Simulation tests

### Nondeterminism test

Nondeterminism test runs a full application simulation, and produces multiple blocks as per the config. It checks the determinism of the application by comparing the apphash at the end of each run to other runs.

The test certifies that, for the same set of operations ( irrespective of what the operations are ), we would reach the same final state if the initial state is the same.

```bash
make test-sim-nondeterminism
```

### Full application simulation test

Full application runs a full app simulation test with the provided configuration.
At the end of the run it tries to export the genesis state to make sure the export works.

```bash
make test-sim-full-app
```

### Multi seed long test

Multi seed long test runs a full application simulation with multiple seeds and multiple blocks. 

This runs the test for a longer duration compared to the multi seed short test.

```bash
make test-sim-multi-seed-long
```

### Multi seed short test

Multi seed short test runs a full application simulation with multiple seeds and multiple blocks.

This runs the test for a longer duration compared to the multi seed long test.

```bash
make test-sim-multi-seed-short
```
---END

**Review by fbac — APPROVED** [review]
LGTM, left some minor changes regarding formatting in a couple markdowns.
